chainer/dataset/convert.py

```diff
@@ -9,6 +10,74 @@
 from chainer import utils
 
 
+class Converter(object):
+    """Base class of batch converters.
+
+    A converter is a callable that takes two arguments: a batch to convert, and
+    a device to send the batch to.
+
+    There are two ways to define a custom converter: to inherit from this class
+    and override :meth:`~chainer.dataset.Converter.converter` method, or to use
+    the :func:`chainer.dataset.converter` decorator.
+    """
+
+    def convert(self, batch, device):
+        """Implements conversion.
+
+        Args:
+            batch: A batch. Value and type are not specified: it depends on the
+                caller of the converter.
+            device (chainer.Device): Device to which the batch is expected to
+                be sent.
+
+        Returns:
+            :Converted batch. Value and type are not specified: it depends on
+            the destination of the converted batch.
+        """
+        raise NotImplementedError(
+            'Converter implementation must override this method.')
+
+    def __call__(self, batch, device):
+        """Invokes the converter.
+
+        This method simply wraps :meth:`chainer.dataset.Converter.converter`.
+        """
+        return self.convert(batch, device)
+
+
+def converter():
+    """Decorator to make a converter function into a \
+:class:`~chainer.dataset.Converter` instance.
+
+    .. admonition:: Example
+
+        >>> @chainer.dataset.converter()
+        >>> def custom_converter(batch, device):
+        >>>     # do something with batch...
+        >>>     return device.send(batch)
+    """
+
+    def wrap(func):
+        @functools.wraps(func)
+        def wrapped_converter(self, batch, device):
+            return func(batch, device)
+
+        @functools.wraps(func)
+        def wrapped_call(self, *args, **kwargs):
+            return func(*args, **kwargs)
+
+        cls = type(
+            func.__name__,
+            (Converter,),
+            {
+                'convert': wrapped_converter,
+                '__call__': wrapped_call,
+            })
+        return cls()
```

Comment:
I guess that the docstring is not appropriately shown with the above code. I feel it's a bit tricky to wrap a function with a callable object which itself is not a function. How about returning a function with a special attribute to distinguish it from a non-decorated one?

Reply:
Hmm, as for the latter part, I think it's not so uncommon to convert a function to non-function with a decorator (`@cupy.fuse` and `@numba.jit` are such examples). What's going on under the hood will be clearer for users. Also, the check (`isinstance(convert, Converter)`) is much easier.

Reply:
As for the compatibility of `concat_examples`, I now think it's enough to document it in the release notes and the upgrade guide. So then, what we need now is a good document rendering.

Reply:
We can render it correctly as a function, by using a custom documenter feature in sphinx.

https://github.com/sphinx-doc/sphinx/blob/d8b69c7f98c1dade6ae1e6b1ddb35d323cf177cd/tests/test_autodoc.py#L524

Reply:
I've made a commit to enable that.
(Currently for some reason the first argument `batch` is omitted from the generated signature. We can override it manually though.)

Reply:
(Sorry for late reply) The documenter feature looks useful for us. Is it needed for users as well when they want to correctly render the documentation for their own converters? That would be better avoided.

Reply:
After offline discussion, I removed `Converter` class and introduce an internal flag to indicate new-style (decorated) converter.